### PR TITLE
Plot neural responses

### PR DESCRIPTION
**Pull request by fkang-pu**
**Summary:**

- Created media directory for storing atlas file. Used "mock.pickle" for now.

- Added a form (NeuronInputParamForm) in form.py with validation code to handle user input. Used hidden fields for both stim_neu_id and resp_neu_ids. 
 
- Added code in view.py to get validated values for form fields, and then call wormfunconn functions to generate data for plotting neural responses.
 
- New template file (plot_neural_responses.html) for form input and output.
 
- Minor changes to navbar.html

**Notes:**

- Ran Superlinter locally and passed. No new test code, although user input parameters can be easily verified as they are listed with graph in output section.

- Tried to get main steps of the web app. in place. Used "vanillaSelectBox" and form hidden fields for selecting neurons. I am open to suggestions or a different approach.
 
- Keep list_input_params.html and related code for now (as references for testing purpose). Plan to delete them later.
 
- Need suggestions on how to handle exceptions while calling wormfunconn functions. I used "raised ValueError..." in views.py, but not sure if it's the right way to handle it. The current version of wormfunconn only worked for **stim_type=rectangular.**

### CODE REVIEW

**Review by lparsons — CHANGES_REQUESTED** [review]
This looks like a solid step forward, nice job Fan!

My one concern is the lack of tests, at least [tests that views return valid pages (200)](https://docs.djangoproject.com/en/3.2/intro/tutorial05/#test-a-view) and a simple request returns a plot (200).

As for the exception handling, I don't see there being much use to catching and re-raising Exceptions. With DEBUG on, we should see the exceptions, and with DEBUG off, they should be hidden. At some point, it would probably be good to report the errors on production in some way. Django docs on [error reporting](https://docs.djangoproject.com/en/3.2/howto/error-reporting/).

**Comment by fkang-pu** [comment]
I am adding some test code, and will push new commits next Monday. If you haven't reviewed this PR, it's better to wait till next Monday.

**Comment by fkang-pu** [comment]
Just pushed a commit. Added error catch for generating html output ((plot_div) from **ployly.offline.plot** function in views.py. Since **plot** function runs validation and including plot.js in output as default, the html output should render a graph is no error is raised.  Also modified and added code to test.py - check form status and expected output or errors.

Ran Super-linter locally and passed. However, GitHub Superlinter failed with the error: Docker pull failed with exit code 1.

**Comment by lparsons** [comment]
@fkang and @jcmatese I believe the linting failure was just a temporary issue, however, I noticed that we were still on v3 of superlinter and decided to take the opportunity to bump to v4 and use the "slim" image which might help speed things up. If you use that when running locally, I've noticed it does speed things up a bit. I'm taking a look at the PR now.

**Review by lparsons — APPROVED** [review]
The tests look good to me, nice work Fan. I believe that linter error was just temporary. I did submit a PR to this branch to update to v4 and use the slim version of Super Linter. Optional if you want to include that or not.

**Review by jcmatese — APPROVED** [review]

**Comment by lparsons** [comment]
@fkang and @jcmatese : While I'm always happy to have feedback from @hepcat72, I don't want to overburden people with reviews. Since Rob is already working on https://github.com/PrincetonUniversity/neuron-graph, we can let him off the hook for reviewing these PRs (though your welcome to if you feel you have the time Rob).

**Comment by fkang-pu** [comment]
It's totally fine with me. I'll let Rob to decide if he wants to review this PR.  I can wait till tomorrow morning.

**Comment by hepcat72** [comment]
I'm fine reviewing. Just wrapped up the updates to the TB data submission PR.  If in the future, I'm slow to review, I wouldn't be offended if you merge without my review.  Or wait, or poke-tag me.  Whatever you want to do is fine by me.

File: neuronsimulator/forms.py
Comment (hepcat72) [inline neuronsimulator/forms.py:53]: NB.
```suggestion
                    "Response neuron(s) cannot include stimulated neuron(s), please make a different selection."
```

File: neuronsimulator/forms.py
Comment (hepcat72) [inline neuronsimulator/forms.py:61]: NB.
```suggestion
                "t_max must be larger than dur."
```

File: neuronsimulator/forms.py
Comment (hepcat72) [inline neuronsimulator/forms.py:82]: NB.
```suggestion
                "You must select at least one response neuron (for testing)."
```

File: neuronsimulator/forms.py
Comment (hepcat72) [inline neuronsimulator/forms.py:90]: NB. Given multiple possibilities, I recommend collection all the errors first and report which ones are invalid.

```
            inv_neurs = []
            for resp_neuron in resp_neu_arr:
                if resp_neuron not in neurons:
                    inv_neurs.append(resp_neuron)
            if len(inv_neurs) > 0:
                raise forms.ValidationError(
                    f"Invalid response neuron(s) encountered: [{', '.join(inv_neurs)}]"
                )
```

File: neuronsimulator/templates/home.html
Comment (hepcat72) [inline neuronsimulator/templates/home.html:8]: NB. I try and avoid placeholders like this, but if you prefer to use them, I suggest employing one or more these tricks I have used in the past to easily find them in the future:
- Comment out the code - or put separate comments above and below.
- Use a keyword that you can use to find placeholders later, e.g. `#PLACEHOLDER#` or `#TODO#`.
- Create a github issue that either/both links to the line of code or add a link to the issue in the comment in the code.

File: neuronsimulator/templates/neuronsimulator/plot_neural_responses.html
Comment (hepcat72) [inline neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:43]: Q. Why are these select lists not located inside `form` tags? Are the form tags in a parent template or are these used in javascript?

File: neuronsimulator/templates/neuronsimulator/plot_neural_responses.html
Comment (hepcat72) [inline neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:74]: C. I often use `<small>` tags like this, and have gotten feedback that it should be done with CSS.

File: neuronsimulator/templates/neuronsimulator/plot_neural_responses.html
Comment (hepcat72) [inline neuronsimulator/templates/neuronsimulator/plot_neural_responses.html:133]: NB. Remove commented debug print.

File: neuronsimulator/views.py
Comment (hepcat72) [inline neuronsimulator/views.py:42]: NB. Remove commented debug prints.

File: neuronsimulator/views.py
Comment (hepcat72) [inline neuronsimulator/views.py:56]: NB. I know that the form has a min value of 1 in the form validation, but for security concerns (because client-side validation can be circumvented), I think doing a server-side code check that `nt` is greater than 0 would prevent an exception upon invalid input.

File: neuronsimulator/views.py
Comment (hepcat72) [inline neuronsimulator/views.py:66]: NB. Remove commented debug print.

Comment (hepcat72) [reply neuronsimulator/views.py:66]: here and elsewhere (e.g. line 76)

Comment (hepcat72) [reply neuronsimulator/views.py:66]: ... or wrap them in a conditional on the value of `settings.DEBUG`.

**Review by hepcat72 — APPROVED** [review]
This is great work Fan.  Feel free to ignore all my knits.  Approved.

**Comment by fkang-pu** [comment]
@hepcat72 Rob, thanks a lot for you detailed review and suggestions, especially the code for invalid reps_neurons.  I will make changes accordingly. Just address your main comments here instead of one-by-one: 

- **form filed names and descriptions**: I used those listed in Francesco's` wormfunconn` package. I won't make changes until I get feedback from him. His code can only handle one stim_type right now, so I expect more changes will be made on his end.

- **debug code**: plan to do a final clean up when we are close to move the code to production. In the next stage of development/testing, I'll still need those lines of debug code.

- **stim_neu_id and resp_neu_ids selections**: I want to use JavaScript code from VanillaSelectBox (a nice web front for multiple selections with auto suggestion feature). I tried different options and also had a discussion with John. The approach I used is the best I can come up with right now.

**Comment by fkang-pu** [comment]
@lparsons I made some minor changes based on Rob's suggestions. Tested again with test.py and from web front. I had to fetch/merge the changes from the remote branch (from your PR7 initially) to my local branch before pushing new commits to GitHub. You may want to take a quick look before I merge this PR to main. Thanks.

File: neuronsimulator/forms.py
Comment (lparsons) [inline neuronsimulator/forms.py]: Just a couple of [options that are a bit more "pythonic"](https://www.geeksforgeeks.org/python-difference-two-lists/).

Use set difference operation:
```python
invalid_neurons = list( set(resp_neu_arr) - set(neurons) )
```

Using list comprehension:
```python
invalid_neurons = [i for i in resp_neu_arr if i not in neurons]
```

**Review by lparsons — COMMENTED** [review]
Looks great @fkang-pu. I had one suggestion for how you could simplify the code a bit using a more ["pythonic" style](https://www.udacity.com/blog/2020/09/what-is-pythonic-style.html).

File: neuronsimulator/forms.py
Comment (fkang-pu) [reply neuronsimulator/forms.py]: @lparsons Nice suggestions., thanks. I tested both in Django shell and opt to use 2nd one. Just pushed another commit.

**Pull request merged**
